### PR TITLE
Card Picker: Entity Picker

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

778ad9417957c8db7a9cd570dab6b854

## Type of change

- [x] New feature (thank you!)

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request: TODO

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/editor/card-editor/hui-card-picker.ts [hunk 1]
```diff
@@ -94,45 +110,83 @@ export class HuiCardPicker extends LitElement {
...
-        style=${styleMap({
-          width: this._width ? `${this._width}px` : "auto",
-          height: this._height ? `${this._height}px` : "auto",
-        })}
+      <mwc-tab-bar
+        .activeIndex=${this._tabIndex}
+        @MDCTabBar:activated=${(tabIndex) => {
+          this._tabIndex = tabIndex.detail.index;
+        }}
       >
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-card-picker.ts:118]: The tabs should be part of the header of the dialog, so they don't scroll with the content.

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 2]
```diff
@@ -247,6 +256,13 @@ export class HuiDialogEditCard extends LitElement implements HassDialog {
                 </mwc-button>
               `
             : ``}
+          ${this._selectedEntities.length
+            ? html`
+                <mwc-button @click=${this._suggestCards}>
+                  ${this.hass!.localize("ui.common.continue")}
+                </mwc-button>
+              `
+            : ""}
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:265]: What happens when you select entities and then switch back to the by cards tab and select a card?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 3]
```diff
@@ -150,31 +160,56 @@ export class HuiDialogEditCard extends LitElement implements HassDialog {
...
-                rel="noreferrer"
-                dir=${computeRTLDirection(this.hass)}
-              >
-                <mwc-icon-button>
-                  <ha-svg-icon path=${mdiHelpCircle}></ha-svg-icon>
-                </mwc-icon-button>
-              </a>
-            `
-          : ""}`}
+        .heading=${html`
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:163]: You should use the `heading` slot of `ha-dialog` <https://github.com/home-assistant/frontend/blob/dev/src/dialogs/more-info/ha-more-info-dialog.ts#L96>

Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:163]: Or with a tab bar:
<https://github.com/home-assistant/frontend/blob/dev/src/panels/config/entities/dialog-entity-editor.ts#L95-L147>

**Comment by bramkragten** [comment]
I think we should split `create-card` into it's own dialog. It is getting messy in `dialog-edit-card` now.

**Comment by zsarnett** [comment]
I agree this is pretty crowded. Would you like me to do that in this PR or different one?

**Comment by bramkragten** [comment]
This PR please

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        <div slot="primaryAction">
+          <mwc-button @click=${this._cancel}>
+            ${this.hass!.localize("ui.common.cancel")}
+          </mwc-button>
+          ${this._currTabIndex === 1
+            ? html`
+                <mwc-button @click=${this._suggestCards}>
+                  ${this.hass!.localize("ui.common.continue")}
+                </mwc-button>
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:130]: disable when not entities picked?

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  private _handleCardPicked(ev) {
+    const config = ev.detail.config;
+    if (this._params!.entities && this._params!.entities.length) {
+      if (Object.keys(config).includes("entities")) {
+        config.entities = this._params!.entities;
+      } else if (Object.keys(config).includes("entity")) {
+        config.entity = this._params!.entities[0];
+      }
+    }
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:199]: Isn't this logic in the edit card dialog? As it is also used for the suggest card?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 6]
```diff
@@ -47,7 +52,7 @@ declare global {
 
 @customElement("hui-dialog-edit-card")
 export class HuiDialogEditCard extends LitElement implements HassDialog {
-  @property() protected hass!: HomeAssistant;
+  @property({ attribute: false }) protected hass!: HomeAssistant;
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:55]:
```suggestion
  @property({ attribute: false }) public hass!: HomeAssistant;
```

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 7]
```diff
@@ -247,6 +248,13 @@ export class HuiDialogEditCard extends LitElement implements HassDialog {
                 </mwc-button>
               `
             : ``}
+          ${this._selectedEntities.length
+            ? html`
+                <mwc-button @click=${this._suggestCards}>
+                  ${this.hass!.localize("ui.common.continue")}
+                </mwc-button>
+              `
+            : ""}
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:257]: Not used anymore?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 8]
```diff
@@ -69,6 +74,8 @@ export class HuiDialogEditCard extends LitElement implements HassDialog {
 
   @internalProperty() private _dirty = false;
 
+  @internalProperty() private _selectedEntities: string[] = [];
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:77]: Not used?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 9]
```diff
@@ -256,6 +264,105 @@ export class HuiDialogEditCard extends LitElement implements HassDialog {
...
+  private _suggestCards(): void {
+    showSuggestCardDialog(this, {
+      lovelaceConfig: this._params!.lovelaceConfig,
+      saveConfig: this._params!.saveConfig,
+      path: this._params!.path as [number],
+      entities: this._selectedEntities,
+    });
+
+    this.closeDialog();
+  }
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:364]: Not used anymore?

File: src/panels/lovelace/editor/card-editor/hui-entity-picker.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+  protected render(): TemplateResult {
+    return html`
+      <div class="container">
+        <ha-data-table
+          auto-height
+          selectable
+          .id=${"entity_id"}
+          .columns=${this._columns()}
+          .data=${this._allEntities(this.hass).map((entity) => {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-entity-picker.ts:39]: We are still mapping and creating a new object on every update of hass, add that the memoize too, so it just returns the data object.

File: src/panels/lovelace/editor/card-editor/show-create-card-dialog.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+export interface ShowCardDialogParams {
+  lovelaceConfig: LovelaceConfig;
+  saveConfig: (config: LovelaceConfig) => void;
+  path: [number] | [number, number];
+  entities?: string[]; // We can pass entity id's that will be added to the config when a card is picked
+}
+
+const importCreateCardDialog = () =>
+  import(
+    /* webpackChunkName: "hui-dialog-edit-card" */ "./hui-dialog-create-card"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/show-create-card-dialog.ts:13]:
```suggestion
    /* webpackChunkName: "hui-dialog-create-card" */ "./hui-dialog-create-card"
```

File: src/panels/lovelace/editor/card-editor/show-create-card-dialog.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { fireEvent } from "../../../../common/dom/fire_event";
+import { LovelaceConfig } from "../../../../data/lovelace";
+
+export interface ShowCardDialogParams {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/show-create-card-dialog.ts:4]:
```suggestion
export interface CreateCardDialogParams {
```

**Review by bramkragten — COMMENTED** [review]
Much cleaner!

File: src/panels/lovelace/editor/card-editor/hui-entity-picker.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+import type { HomeAssistant } from "../../../../types";
+import type {
+  DataTableRowData,
+  SelectionChangedEvent,
+} from "../../../../components/data-table/ha-data-table";
+
+import "../../../../components/data-table/ha-data-table";
+import "../../../../components/entity/state-badge";
+
+@customElement("hui-entity-picker")
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-entity-picker.ts:27]: Can we name this
```suggestion
@customElement("hui-entity-picker-table")
```

Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-entity-picker.ts:27]: And can we share it between the unused entities view?

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts (hunk 5 above)
Comment (zsarnett) [reply src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:199]: I dont know what you mean by this

File: src/panels/lovelace/editor/card-editor/hui-entity-picker-table.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+                `
+              : ""}
+          </div>
+        `,
+      },
+    };
+
+    if (narrow) {
+      return columns;
+    }
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-entity-picker-table.ts:91]: We should use the new `hidden` options so we can still filter it on mobile

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts [hunk 15]
```diff
@@ -113,17 +117,31 @@ export class HuiCreateDialogCard extends LitElement implements HassDialog {
...
+                    const stateObj = this.hass.states[entity];
+                    return {
+                      icon: "",
+                      entity_id: entity,
+                      stateObj,
+                      name: computeStateName(stateObj),
+                      domain: computeDomain(entity),
+                      last_changed: stateObj!.last_changed,
+                    };
+                  }) as DataTableRowData[]}
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:134]:
```suggestion
                  .entities=${this._allEntities(this.hass.entities)}
```

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts [hunk 16]
```diff
@@ -239,6 +265,8 @@ export class HuiCreateDialogCard extends LitElement implements HassDialog {
 
     this.closeDialog();
   }
+
+  private _allEntities = memoize(() => Object.keys(this.hass.states));
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:269]:
```suggestion
  private _allEntities = memoize((entities) => Object.keys(entities).map((entity) => {
                    const stateObj = this.hass.states[entity];
                    return {
                      icon: "",
                      entity_id: entity,
                      stateObj,
                      name: computeStateName(stateObj),
                      domain: computeDomain(entity),
                      last_changed: stateObj!.last_changed,
                    };
                  }));
```

File: src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts (hunk 5 above)
Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:225]: You can just add `entities` to the options of `showEditCardDialog`.

Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:225]: Oh wait... `create-card-dialog` will open `suggest-card-dialog` and that will open `create-card-dialog` again and this logic kicks in?

Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-create-card.ts:225]: But I don't see that in the code now, suggest card will open `dialog-edit-card`? Does it even work now?

**Review by bramkragten — APPROVED** [review]

**Pull request merged**